File: Nikhil/java/src/main/resources/python/file_download_tool.py
from __future__ import annotations

"""
file_download_tool.py

Python translation of `com.infosys.small.pnbc.FileDownloadTool`.

This tool exposes files uploaded through CAPT and caches them in the
current ExecutionContext (per customer number and document type).
"""

import logging
from typing import Any, Mapping

from pydantic import BaseModel, Field

from api import Api
from chat_types import ToolCall, ToolCallResult
from react_agent import ReactAgent
from scenario_component import ScenarioComponent

# NOTE: This module is expected to be available in the project. Do NOT provide fallbacks.
from capt import Capt  # noqa: F401

# ------------------------------------------------------------------------------
# Logging (equivalent to Java SimpleLogger)
# ------------------------------------------------------------------------------
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger(__name__)


class FileDownloadTool(Api):
    """
    Provides access to files that have been uploaded through CAPT, such as:
    - Proforma Document
    - Power of Attorney (PoA)
    - Probate Certificate (SKS)
    """

    # ---------------------------- parameters ---------------------------- #
    class Parameters(ReactAgent.Parameters):
        """
        Parameters schema mirroring the Java class, with JSON aliases matching the
        original @JsonProperty names.
        """

        customer_number: str = Field(
            ...,
            alias="customerNumber",
            description="Unique customer number of the estate the document refers to.",
        )
        document_type: "Capt.Parameters.DocumentType" = Field(  # type: ignore[name-defined]
            ...,
            alias="documentType",
            description="The type of the document to download.",
        )

        # accept both alias and field name at runtime
        model_config = {"populate_by_name": True}

    # ---------------------------- construction -------------------------- #
    def __init__(self) -> None:
        super().__init__(
            id_="fileDownload",
            description=(
                "This tool allows downloading files like Proforma Document, "
                "Power of Attorney document (PoA) and Probate Certificate (SKS), if available."
            ),
            schema=FileDownloadTool.Parameters,
        )

    # ------------------------------ invoke ------------------------------ #
    def invoke(self, call: ToolCall) -> ToolCallResult:  # noqa: D401
        """
        Execute the file retrieval, caching per (customer_number, document_type).

        Java semantics mirrored:
        * Ensure tool is initialized.
        * Remove LLM-only arguments (e.g., thought).
        * Validate required inputs.
        * Use the ExecutionContext caches (SKS / PoA / Proforma).
        * If first call for the given (customer_number, type), fetch from ScenarioComponent.
        """
        if call is None:  # @NonNull translation
            raise ValueError("call must not be None")
        if not self.is_initialized():
            # Java: IllegalArgumentException → Python: ValueError
            raise ValueError("Tool must be initialized.")

        # Clone & sanitise arguments
        args: dict[str, Any] = dict(call.arguments)
        args.pop("thought", None)  # remove LLM internal field, if present

        # Required parameters (names follow the JSON alias used in Java)
        customer_number = self.get_string("customerNumber", args)
        if customer_number is None:
            return ToolCallResult.from_call(
                call,
                "ERROR: You must provide Customer Number of the estate the file refers to.",
            )

        document_type = self.get_string("documentType", args)
        # Do not pass documentType to underlying scenario tools (match Java)
        args.pop("documentType", None)

        # Access execution context caches
        ctx = self.get_execution_context()
        if ctx is None:
            return ToolCallResult.from_call(call, "ERROR: Execution context is missing.")

        # Route by document type
        if document_type == "SKS":
            if ctx.sks.get(customer_number) is None:
                ctx.sks[customer_number] = ScenarioComponent.get_instance().get(
                    self.get_scenario_id(),
                    "getSKS",
                    args,
                )
            return ToolCallResult.from_call(call, ctx.sks.get(customer_number))

        if document_type == "POWER_OF_ATTORNEY":
            if ctx.poa.get(customer_number) is None:
                ctx.poa[customer_number] = ScenarioComponent.get_instance().get(
                    self.get_scenario_id(),
                    "getPoA",
                    args,
                )
            return ToolCallResult.from_call(call, ctx.poa.get(customer_number))

        if document_type == "PROFORMA_DOCUMENT":
            if ctx.proforma_document.get(customer_number) is None:
                ctx.proforma_document[customer_number] = ScenarioComponent.get_instance().get(
                    self.get_scenario_id(),
                    "getProformaDocument",
                    args,
                )
            return ToolCallResult.from_call(
                call, ctx.proforma_document.get(customer_number)
            )

        # Fallback: invalid type
        return ToolCallResult.from_call(
            call, f"ERROR: Invalid document type: {document_type}"
        )

File: Nikhil/java/src/main/java/com/infosys/small/pnbc/FileDownloadTool.java
/**
 * 
 */
package com.infosys.small.pnbc;

import java.util.HashMap;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.infosys.small.core.ToolCall;
import com.infosys.small.core.ToolCallResult;
import com.infosys.small.pnbc.Capt.Parameters.DocumentType;
import com.infosys.small.react.ReactAgent;

import lombok.NonNull;

/**
 * This tool gives you access to files that have been uploaded through CAPT.
 * This is because asking agents to provide files, sometimes results in a file
 * summary.
 */
public class FileDownloadTool extends Api {

	public static class Parameters extends ReactAgent.Parameters {

		@JsonProperty(required = true)
		@JsonPropertyDescription("Unique customer number of the estate the document refers to.")
		public String customerNumber;

		@JsonProperty(required = true)
		@JsonPropertyDescription("The type of the document to download.")
		public DocumentType documentType;
	}

	@Override
	public ToolCallResult invoke(@NonNull ToolCall call) throws Exception {
		if (!isInitialized()) {
			throw new IllegalArgumentException("Tool must be initialized.");
		}

		Map<String, Object> args = new HashMap<>(call.getArguments());
		args.remove("thought"); // As we are passing it, we must remove or it won't match tools

		String customerNumber = getString("customerNumber", args);
		if (customerNumber == null)
			return new ToolCallResult(call,
					"ERROR: You must provide Customer Number of the estate the file refers to.");

		String documentType = getString("documentType", args);
		args.remove("documentType"); // As we are passing it to tools, we must remove or it won't match tools
		switch (documentType) {
		case "SKS":
			if (getExecutionContext().getSKS().get(customerNumber) == null) {
				// This is the first time we call this API; let's get value from scenario
				getExecutionContext().getSKS().put(customerNumber, ScenarioComponent.getInstance().get( //
						getScenarioId(), //
						"getSKS", //
						args));
			}
			return new ToolCallResult(call, getExecutionContext().getSKS().get(customerNumber));
		case "POWER_OF_ATTORNEY":
			if (getExecutionContext().getPoA().get(customerNumber) == null) {
				// This is the first time we call this API; let's get value from scenario
				getExecutionContext().getPoA().put(customerNumber, ScenarioComponent.getInstance().get( //
						getScenarioId(), //
						"getPoA", //
						args));
			}
			return new ToolCallResult(call, getExecutionContext().getPoA().get(customerNumber));
		case "PROFORMA_DOCUMENT":
			if (getExecutionContext().getProformaDocument().get(customerNumber) == null) {
				// This is the first time we call this API; let's get value from scenario
				getExecutionContext().getProformaDocument().put(customerNumber, ScenarioComponent.getInstance().get( //
						getScenarioId(), //
						"getProformaDocument", //
						args));
			}
			return new ToolCallResult(call, getExecutionContext().getProformaDocument().get(customerNumber));
		default:
			return new ToolCallResult(call, "ERROR: Invalid document type: " + documentType);
		}
	}

	public FileDownloadTool() {

		super("fileDownload", //
				"This tool provides access to files stored on backend system such as Proforma Document, Power of Attorney document (PoA), and Probate Certificate (SKS).",
				Parameters.class);
	}
}